model-mesh-ovms-adapter/server/modelmanager.go

```diff
@@ -0,0 +1,499 @@
+		// requests; here we take the approach of waiting for a fixed
+		// period of time after a request is received
+		batch := make([]*request, 0, 10)
+		var stopChan <-chan time.Time = nil
+
+	runLoopSelect:
+		select {
+		case req, ok := <-mm.requests:
+			if !ok {
+				// shutdown the loop after processing this batch
+				mm.requests = nil
+				break // from select
+			}
+			// add request to batch
+			batch = append(batch, req)
+
+			// set the stop timer, if not set already
+			if stopChan == nil {
+				stopChan = time.NewTimer(100 * time.Millisecond).C
+			}
+
+			goto runLoopSelect
+		case <-stopChan:
+			break // from select
+		}
+
+		log.V(1).Info("Processing batch of requests", "numRequests", len(batch))
+
+		// update the model configuration based on the requests
+		var unloadAllRequest *request
+
+		// build a new map to track changes to the same model
+		modelUpdates := map[string]*request{}
+		for _, req := range batch {
+			// check for an unloadAll
+			if req.requestType == unloadAll {
+				unloadAllRequest = req
+				// abort any requests preceding the unloadAll
+				for id, req := range modelUpdates {
+					completeRequest(req, codes.Aborted, "Model Server is being reset")
+					// remove the pointer from the updates map
+					delete(modelUpdates, id)
+				}
+				continue
+			}
+
+			if modelUpdates[req.modelId] != nil {
+				// abort the prior request first
+				completeRequest(modelUpdates[req.modelId], codes.Aborted, "Concurrent request received for the same model")
+			}
+
+			modelUpdates[req.modelId] = req
+		}
+
+		// if the batch included an UnloadAll, reset the model config map prior adding the updates
+		if unloadAllRequest != nil {
+			mm.loadedModelsMap = map[string]OvmsMultiModelConfigListEntry{}
+		}
+
+		// process the updates
+		for id, req := range modelUpdates {
+			switch req.requestType {
+			case load:
+				mm.loadedModelsMap[id] = OvmsMultiModelConfigListEntry{
+					Config: OvmsMultiModelModelConfig{
+						Name:     req.modelId,
+						BasePath: req.basePath,
+					},
+				}
+			case unload:
+				delete(mm.loadedModelsMap, id)
+			}
+		}
+
+		// reload the config
+		ctx := context.TODO()
+		if err := mm.updateModelConfig(ctx); err != nil {
+			msg := "Failed to update model configuration with OVMS"
+			log.Error(err, msg)
+
+			// at this point, we don't know whether OVMS has
+			// reloaded or not... so complete all requests with
+			// errors
+
+			for _, req := range modelUpdates {
+				completeRequest(req, codes.Internal, msg)
+			}
+
+			if unloadAllRequest != nil {
+				completeRequest(unloadAllRequest, codes.Internal, msg)
+			}
+
+			continue // back to the start of the main loop
+		}
+
+		// complete the requests
+		for id, req := range modelUpdates {
+			var statusExists bool
+			var modelStatus OvmsModelVersionStatus
+
+			conf, statusExists := mm.cachedModelConfigResponse[id]
+			if statusExists {
+				modelStatus = conf.ModelVersionStatus[0]
+			}
+
```

Comment:
I think this is where things get a bit hairy with the distributed state we are dealing with.
What are the assumed semantics of a failure for a load/unload request in ModelMesh? If a model unload fails does it assume that the model is loaded and vice-versa? I know the answer is "no" since it will send an Unload after a failed Load.

Maybe `loadedModelsMap` is mis-named. I am treating it as the _desired_ configuration instead of the actual configuration (similar to how the multi-model config is managed by OVMS). LoadModel/UnloadModel are changing the desired state and asking for the current state. Is that an ok way to think about the requests coming from ModelMesh?

Reply:
> What are the assumed semantics of a failure for a load/unload request in ModelMesh? If a model unload fails does it assume that the model is loaded and vice-versa? I know the answer is "no" since it will send an Unload after a failed Load.

Model-mesh acts defensively in this situation which is why it sends the unload if a load fails, but I don't think the adapter should rely on that and should aim to provide its own clean semantics / defensive behaviour i.e. if model load fails then it's as if that request wasn't made.

FWIW the behaviour of model-mesh after a failed _unload_ is to retry periodically and after some time it gives up and considers that model's space as "lost" (yes probably the runtime is in a bad state if this happens so may be best to trigger a controlled restart somehow).

> The state of a model is not really known unless a request succeeds.

Again I agree this is true from model-mesh's pov but I feel its better to not assume/rely on that fact from the adapter side.


The other advantage of doing this pruning is that we can always act directly on the map per the other comment.

> LoadModel/UnloadModel are changing the desired state and asking for the current state. Is that an ok way to think about the requests coming from ModelMesh?

Yes, though I'm not sure whether I know what you mean by "asking for" here...